### PR TITLE
Add 2.6 prebuilt models to Machine Learning

### LINKED ISSUE #2565: [DOC] [ml-commons] Add prebuilt model to doc
**What do you want to do?**
 
- [x] Request a change to existing documentation
- [ ] Add new documentation
- [ ] Report a technical problem with the documentation
- [ ] Other 

**Tell us about your request.** Provide a summary of the request and all versions that are affected.

We have built some popular text embedding models and user can use these models easily by specifying model name and version only. No effort needed to trace model and prepare model parameters. We should list these models in a table and add to ml-commons doc.

**What other resources are available?** Provide links to related issues, POCs, steps for testing, etc.

### PR DESCRIPTION
**Pull request by Naarcha-AWS**
Fixes #2565 

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ylwu-amzn** [comment]
You can find all models in this file https://github.com/opensearch-project/ml-commons/blob/2.x/ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_embedding/pre_trained_models_list.json

File: _ml-commons-plugin/prebuilt-models.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+---
+layout: default
+title: Pre-trained models
```
Comment (JeffH-AWS) [inline _ml-commons-plugin/prebuilt-models.md:3]: This is confusing - are they pre-built models as the name of the doc implies, or are they pre-trained models like the title and content implies? They should be consistent.

File: _ml-commons-plugin/prebuilt-models.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+parent: Model-serving framework
+nav_order: 120
+---
+
```
Comment (JeffH-AWS) [inline _ml-commons-plugin/prebuilt-models.md:7]: The page is missing an h1 title.

File: _ml-commons-plugin/prebuilt-models.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+The model-serving framework supports a variety of open-source pre-built models that can assist with a variety of machine learning (ML) search and analytics use-cases.
```
Comment (JeffH-AWS) [inline _ml-commons-plugin/prebuilt-models.md:8]: Here the term "pre-built" is used again instead of "pre-trained"

File: _ml-commons-plugin/prebuilt-models.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+To use a pre-trained model in your OpenSearch cluster:
+
+1. Select the model and download the model artifact and config.json file. For a list of pre-trained models, see [supported pre-trained model](#supported-pre-trained-models).
+2. Upload the model using the [upload API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-serving-framework#upload-model-to-opensearch). For more information on how to upload and use model, see [Model-serving framework]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-serving-framework).
+
+## Supported pre-trained models
+
+The model-serving framework supports the following models. 
+
+Although models might have similar use cases, they many of them contain different performance metrics and model sizes. For a comparison of the performances of pre-trained models, see the [sbert documentation](https://www.sbert.net/docs/pretrained_models.html#model-overview).
```
Comment (JeffH-AWS) [inline _ml-commons-plugin/prebuilt-models.md:21]: Is there a way to weave this into the content? The previous line says "...supports the following models." but the next object on the page is this warning text, and it's followed by an h3 section that doesn't really tie in to the previous section. Are "Sentence transformers" the "models supported by the model-serving framework"? Or are the rows in the subsequent table the supported models? It isn't clear to me as a reader.

File: _ml-commons-plugin/prebuilt-models.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+{: .note}
+
+
+### Sentence transformers
+
+Sentence transformer models map sentences and paragraphs across a dimensional dense vector space. The number of vectors depends on the model. Use these models for use cases such as clustering and semantic search.
+
+| **Model name** | **Vector dimensions** | **Torchscript artifact** | **ONNX artifact** |
+|---|---|---|---|
+| `sentence-transformers/all-distilroberta-v1` | 768 dimensional dense vector space. | - [model_url](https://artifacts.opensearch.org/models/ml-models/huggingface/sentence-transformers/all-distilroberta-v1/1.0.1/torch_script/sentence-transformers_all-distilroberta-v1-1.0.1-torch_script.zip)<br>- [config_url](https://artifacts.opensearch.org/models/ml-models/huggingface/sentence-transformers/all-distilroberta-v1/1.0.1/torch_script/config.json) | - [model_url](https://artifacts.opensearch.org/models/ml-models/huggingface/sentence-transformers/all-distilroberta-v1/1.0.1/onnx/sentence-transformers_all-distilroberta-v1-1.0.1-onnx.zip)<br>- [config_url](https://artifacts.opensearch.org/models/ml-models/huggingface/sentence-transformers/all-distilroberta-v1/1.0.1/onnx/config.json) |
```
Comment (JeffH-AWS) [inline _ml-commons-plugin/prebuilt-models.md:31]: Is it correct to use periods for the data in column 2? E.g.:

```
768 dimensional dense vector space.
```

That isn't a sentence - should we still use periods?

**Review by JeffH-AWS — DISMISSED** [review]
Left a few comments

File: _ml-commons-plugin/prebuilt-models.md (hunk 4 above)
Comment (Naarcha-AWS) [reply _ml-commons-plugin/prebuilt-models.md:21]: I added a "categorized based on type" qualifier here. I'm breaking a rule I have where I don't like to add single nested levels in headers, but the plan is to add more pre-trained model types (such as text embedders) in each subsequent release. So, as opposed to adding a "Type" column and having a giant table in the future, and to prevent repeating the same description over and over again, we'll add a new table and a new h3 header for each model type.

File: _ml-commons-plugin/prebuilt-models.md (hunk 5 above)
Comment (Naarcha-AWS) [reply _ml-commons-plugin/prebuilt-models.md:31]: I'll defer to Nate Bower on this one.

File: _ml-commons-plugin/pretrained-models.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+## Supported pre-trained models
+
+The model-serving framework supports the following models, categorized based on type. Although models with the same type will have similar use cases, each model has a different model size and performs differently depending on your cluster. For a comparison of the performances some pre-trained models, see the [sbert documentation](https://www.sbert.net/docs/pretrained_models.html#model-overview).
+
+
+### Sentence transformers
+
+Sentence transformer models map sentences and paragraphs across a dimensional dense vector space. The number of vectors depends on the model. Use these models for use cases such as clustering and semantic search.
+
+| **Model name** | **Vector dimensions** | **Torchscript artifact** | **ONNX artifact** |
```
Comment (carolxob) [inline _ml-commons-plugin/pretrained-models.md:28]: Is there a new standard that requires table headers to be bold? Should we be doing this across all docs?

**Review by carolxob — APPROVED** [review]
LGTM with Jeff's feedback and one question.

File: _ml-commons-plugin/pretrained-models.md (hunk 6 above)
Comment (Naarcha-AWS) [reply _ml-commons-plugin/pretrained-models.md:28]: We don't have a standard one way or another. I always bold mine, so that the header labels are more obvious.

Since it's a formatting question, lets get a consensus and add it to the formatting guide.

File: _ml-commons-plugin/pretrained-models.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
+---
+layout: default
+title: Pre-trained models
```
Comment (natebower) [inline _ml-commons-plugin/pretrained-models.md:3]:
```suggestion
title: Pretrained models
```

File: _ml-commons-plugin/pretrained-models.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+parent: Model-serving framework
+nav_order: 120
+---
+
+# Pretrained models
+
+The model-serving framework supports a variety of open-source pretrained models that can assist with a variety of machine learning (ML) search and analytics use cases.
```
Comment (natebower) [inline _ml-commons-plugin/pretrained-models.md:10]:
```suggestion
The model-serving framework supports a variety of open-source pretrained models that can assist with a range of machine learning (ML) search and analytics use cases. 
```

Comment (natebower) [reply _ml-commons-plugin/pretrained-models.md:10]: Avoiding the double use of "variety".

File: _ml-commons-plugin/pretrained-models.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+# Pretrained models
+
+The model-serving framework supports a variety of open-source pretrained models that can assist with a variety of machine learning (ML) search and analytics use cases. 
+
+## Uploading pretrained models
+
+To use a pretrained model in your OpenSearch cluster:
+
+1. Select the model and download the model artifact and config.json file. For a list of pretrained models, see [supported pretrained model](#supported-pretrained-models).
```
Comment (natebower) [inline _ml-commons-plugin/pretrained-models.md:16]: In the link, should "model" be plural?

File: _ml-commons-plugin/pretrained-models.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+2. Upload the model using the [upload API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-serving-framework#upload-model-to-opensearch). For more information on how to upload and use model, see [Model-serving framework]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-serving-framework).
```
Comment (natebower) [inline _ml-commons-plugin/pretrained-models.md:17]:
```suggestion
2. Upload the model using the [upload API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-serving-framework#upload-model-to-opensearch). For more information on how to upload and use ML models, see [Model-serving framework]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-serving-framework).
```

File: _ml-commons-plugin/pretrained-models.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+## Supported pretrained models
+
+The model-serving framework supports the following models, categorized based on type. Although models with the same type will have similar use cases, each model has a different model size and performs differently depending on your cluster. For a comparison of the performances some pre-trained models, see the [sbert documentation](https://www.sbert.net/docs/pretrained_models.html#model-overview).
```
Comment (natebower) [inline _ml-commons-plugin/pretrained-models.md:21]:
```suggestion
The model-serving framework supports the following models, categorized by type. Although models with the same type will have similar use cases, each model has a different model size and performs differently depending on your cluster. For a comparison of the performances of some pretrained models, see the [sbert documentation](https://www.sbert.net/docs/pretrained_models.html#model-overview).
```

File: _ml-commons-plugin/pretrained-models.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+
+### Sentence transformers
+
+Sentence transformer models map sentences and paragraphs across a dimensional dense vector space. The number of vectors depends on the model. Use these models for use cases such as clustering and semantic search.
+
+| **Model name** | **Vector dimensions** | **Torchscript artifact** | **ONNX artifact** |
```
Comment (natebower) [inline _ml-commons-plugin/pretrained-models.md:28]: We should introduce the table in the text that precedes it. "The following table provides..."

File: _ml-commons-plugin/pretrained-models.md (hunk 6 above)
Comment (natebower) [reply _ml-commons-plugin/pretrained-models.md:28]: Agreed. They should generally always be bold.

**Review by natebower — COMMENTED** [review]
@Naarcha-AWS Please see my comments and changes and let me know if you have any questions. Thanks!

File: _ml-commons-plugin/model-serving-framework.md [hunk 13]
```diff
@@ -21,7 +21,7 @@ To upload a custom model to OpenSearch, you need to prepare it outside of your O
 
 ### Model support
 
-As of OpenSearch 2.4, the model-serving framework only supports text embedding models without GPU acceleration.
+As of OpenSearch 2.6, the model-serving framework supports text embedding and sentence transformer models.
```
Comment (ylwu-amzn) [inline _ml-commons-plugin/model-serving-framework.md:24]: "sentence transformer" is just a kind of text embedding model. Let's just use text embedding models here?

**Review by hdhalter — COMMENTED** [review]
@Naarcha-AWS , can you please update the title of the PR to say ML-commons or machine learning since this will go in the release notes? Thanks!

File: _ml-commons-plugin/pretrained-models.md [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+# Pretrained models
+
+The model-serving framework supports a variety of open-source pretrained models that can assist with a range of machine learning (ML) search and analytics use cases. 
+
+## Uploading pretrained models
+
+To use a pretrained model in your OpenSearch cluster:
+
+1. Select the model and download the model artifact and config.json file. For a list of pretrained models, see [supported pretrained models](#supported-pretrained-models).
```
Comment (ylwu-amzn) [inline _ml-commons-plugin/pretrained-models.md:16]: Upload models from our model repo is easy. User just need to provide model name, version and format. The supported model version starts from 1.0.1 

check https://github.com/opensearch-project/ml-commons/issues/745

```
POST /_plugins/_ml/models/_upload
{
  "name": "huggingface/sentence-transformers/all-MiniLM-L12-v2",
  "version": "1.0.1",
  "model_format": "TORCH_SCRIPT"
}
```

**Review by cwillum — APPROVED** [review]

File: _ml-commons-plugin/pretrained-models.md (hunk 14 above)
Comment (dhrubo-os) [inline _ml-commons-plugin/pretrained-models.md:16]: After `Select the model and download the model artifact and config.json file.` can we refer to the section `Sentence transformers`, so that customers don't get confused which models they need to download.

Comment (Naarcha-AWS) [reply _ml-commons-plugin/pretrained-models.md:16]: @dhrubo-os: Since the plan is to add more models in the future, I would rather link to the general section.

File: _ml-commons-plugin/prebuilt-models.md (hunk 1 above)
Comment (dylan-tong-aws) [reply _ml-commons-plugin/prebuilt-models.md]: Agree, about the comment on consistency. The industry standard term is pretrained models.

File: _ml-commons-plugin/pretrained-models.md [hunk 15]
```diff
@@ -0,0 +1,40 @@
+---
+layout: default
+title: Pretrained models
+parent: Model-serving framework
+nav_order: 120
+---
+
+# Pretrained models
+
+The model-serving framework supports a variety of open-source pretrained models that can assist with a range of machine learning (ML) search and analytics use cases.
```
Comment (dhrubo-os) [inline _ml-commons-plugin/pretrained-models.md:10]: May be we can add:

We traced a few pre-trained models in our model repo for customers so that they can use these models in opensearch.

File: _ml-commons-plugin/pretrained-models.md (hunk 14 above)
Comment (Naarcha-AWS) [reply _ml-commons-plugin/pretrained-models.md:16]: How would the API know that we're uploading a pretrained model unless they download the artifact first? Does it come prepackaged with OpenSearch?

Comment (ylwu-amzn) [reply _ml-commons-plugin/pretrained-models.md:19]: When upload model with "Bring your own model(BYOM)" way, user need to provide "url" in request parameter, for example
```
POST /_plugins/_ml/models/_upload
{
  "name": "all-MiniLM-L6-v2",
  "version": "1.0.0",
  "description": "test model",
  "model_format": "TORCH_SCRIPT",
  "model_config": {
    "model_type": "bert",
    "embedding_dimension": 384,
    "framework_type": "sentence_transformers",
    "all_config": "{\"_name_or_path\":\"nreimers/MiniLM-L6-H384-uncased\",\"architectures\":[\"BertModel\"],\"attention_probs_dropout_prob\":0.1,\"gradient_checkpointing\":false,\"hidden_act\":\"gelu\",\"hidden_dropout_prob\":0.1,\"hidden_size\":384,\"initializer_range\":0.02,\"intermediate_size\":1536,\"layer_norm_eps\":1e-12,\"max_position_embeddings\":512,\"model_type\":\"bert\",\"num_attention_heads\":12,\"num_hidden_layers\":6,\"pad_token_id\":0,\"position_embedding_type\":\"absolute\",\"transformers_version\":\"4.8.2\",\"type_vocab_size\":2,\"use_cache\":true,\"vocab_size\":30522}"
  },
  "url": "https://github.com/ylwu-amzn/ml-commons/blob/2.x_custom_m_helper/ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_embedding/all-MiniLM-L6-v2_torchscript_sentence-transformer.zip?raw=true"
}
```
When upload pre-trained model from our model repo, user don't need to provide URL. If no URL provided, our code will try to construct pre-trained model URL with model name, version and format like this way
`https://artifacts.opensearch.org/models/ml-models/<model_name>/<version>/<model format>/<model_artifact_zip>`

If model name is `huggingface/sentence-transformers/msmarco-distilbert-base-tas-b`,  version is `1.0.1`, format is `onnx`, then model zip URL is  `https://artifacts.opensearch.org/models/ml-models/huggingface/sentence-transformers/msmarco-distilbert-base-tas-b/1.0.1/onnx/sentence-transformers_msmarco-distilbert-base-tas-b-1.0.0-torch_script.zip`

File: _ml-commons-plugin/prebuilt-models.md (hunk 5 above)
Comment (Naarcha-AWS) [reply _ml-commons-plugin/prebuilt-models.md]: Added. I linked to the model-serving framework forum.

File: _ml-commons-plugin/pretrained-models.md (hunk 14 above)
Comment (ylwu-amzn) [reply _ml-commons-plugin/pretrained-models.md:19]: For pre-trained model, user done't need to download model zip and config
```suggestion
1. Select the model for a list of pretrained models, see the pre-trained model table below (add table link)
```

The model config json file is not readable, suggest link to the model table below.

File: _ml-commons-plugin/pretrained-models.md (hunk 15 above)
Comment (Naarcha-AWS) [reply _ml-commons-plugin/pretrained-models.md:13]: Addressed this by adding a caveat to the sbert link.
